File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0652_Find_Duplicate_Subtrees.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Question_0652_Find_Duplicate_Subtrees {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        List<TreeNode> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Map<String, List<TreeNode>> map = new HashMap<>();
        serialize(root, map);
        for (Map.Entry<String, List<TreeNode>> entry : map.entrySet()) {
            if (entry.getValue().size() > 1) {
                res.add(entry.getValue().get(0));
            }
        }
        return res;
    }

    private String serialize(TreeNode root, Map<String, List<TreeNode>> map) {
        if (root == null) {
            return "";
        }
        String s = root.val + "(" + serialize(root.left, map) + "," + serialize(root.right, map) + ")";
        if (!map.containsKey(s)) {
            map.put(s, new ArrayList<>());
        }
        map.get(s).add(root);
        return s;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0652_Find_Duplicate_Subtrees.cc
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
    vector<TreeNode*> res;
    if (!root) {
      return res;
    }
    unordered_map<string, vector<TreeNode*>> map;
    serialize(root, map);
    for (auto it = map.begin(); it != map.end(); it++) {
      if (it->second.size() > 1) {
        res.push_back(it->second[0]);
      }
    }
    return res;
  }

 private:
  string serialize(TreeNode* root, unordered_map<string, vector<TreeNode*>>& map) {
    if (!root) {
      return "";
    }
    string s = to_string(root->val) + "(" + serialize(root->left, map) + "," + serialize(root->right, map) + ")";
    map[s].push_back(root);
    return s;
  }
};
